RationaleMCP/0035/Readme.md

```diff
@@ -0,0 +1,129 @@
+* Which strings should be translated:
+
+The precise updated text of the specification is part of this branch/pull-request: Yes.
+
+# Backwards Compatibility
+The proposal is backwards compatible. All multilingual text changes can be applied without effect at existing Modelica code. 
+After the introduction of translation files for libraries, this files need to be adjusted with every library change which adds new commented parts or changes existing comments.
+
+# Tool Implementation
+
+## Experience with Prototype
+### Implementation steps
+#### 1. Read out a translation template file
+Create at first a translation template file \*.pot for a Modelica library. Means a (tool dependent) readout of all needed text strings into a *LibraryName*.pot-File. It contains: 
+- a header which concludes all necessary information
+![Header of a *.pot file](POT_Header.png)
+
+- for all original strings entries with following structure: 
+```
+    msgctxt "Hydraulics.Basics.TankOverflow"
+    msgid "Volume Flow Port A"
+    msgstr ""
+```
+
+Where *msgctxt* contains the name of the Modelica Class and *msgid* contains the textstring which is situated in this name space and shall be translated. Please regard that if a *msgid* string (in this example "Volume Flow Port A") is given more than once in a namespace, all occurrences are translated with the same translation!
+
+It shall be mandatory to use for *msgctxt* the full name of a class:
+- so a specific translation depending on the content in this namespace is possible, as every textstring may differ depending on the context,
+- using the class makes it easier for the Modelica tool to copy and re-arrange models and packages without loosing the already existing language information.
+
+The following Modelica constructs shall be read out / translated: 
+* syntactical comments
+* strings in following annotations:
+  * Text.string, Text.textString
+  * missingInnerMessage, obsolete, unassignedMessage 
+  * Dialog.[group|tab] 
+  * Dialog.[load|save]Selector.[caption|filter] 
+  * Documentation.[info|revisions]
+  * Figure.title, Plot.title, Curve.legend
+
+Having read out the file it just needs to be changed, if the library is changed (e.g. commented parts are added or changed or when having changed the Modelica name of the text strings containing element). 
+
+#### 2. Translation of the template file into a language file
+Starting from the template file create a \*.po file for each needed language (e.g. german: *LibraryName*.de.po-File).
+
+Edit the *msgstr* with the translation in the wanted language (here for german: `` Hydraulics.de.po``) :
+```
+    msgctxt "Hydraulics.Basics.TankOverflow"
+    msgid "Volume Flow Port A"
+    msgstr "Volumenstrom an Anschluss A"
+```
+
+Hereby no error shall occur if there is one comment not having a translation. In this case the not translated text (the content of *msgid*) shall be used.

```

Comment:
```suggestion
Hereby no error shall occur if there is one comment not having a translation, or even is missing from the translation file. In this case the not translated text (the content of *msgid*) shall be used.

```
Error handling is always important in practice.
Clearly it is good to check the translation file to see that nothing is missing, but when using a model such as TankOverflow it seems better to show the original text than just giving an error. (But other alternatives are also possible.)

I also noticed that this was missing from the specification text - but it is part of some gettext API-interfaces and thus can be seen as being implicitly included; although others fail in that case.

Reply:
Ok, may be implicit that missing translation is part of not having a translation.